src/transformers/generation/utils.py

```diff
@@ -4550,62 +4558,41 @@ def assisted_decoding(
             candidate_input_ids = input_ids
-            assistant_attention_mask = model_kwargs.get("attention_mask", None)
-            assistant_decoder_attention_mask = model_kwargs.get("decoder_attention_mask", None)
-            assistant_encoder_outputs = (model_kwargs.get("assistant_encoder_outputs", None),)
+            assistant_kwargs = copy.copy(model_kwargs)
+            # import pdb; pdb.set_trace()
             for _ in range(int(num_assistant_tokens)):
-                # 1.1. use the assistant model to obtain the next candidate logits
+                # 1.1 prepare assistant model inputs
                 assistant_inputs = assistant_model.prepare_inputs_for_generation(
                     candidate_input_ids,
-                    attention_mask=assistant_attention_mask,
-                    decoder_attention_mask=assistant_decoder_attention_mask,
-                    encoder_outputs=assistant_encoder_outputs,
-                    past_key_values=model_kwargs.get("assistant_past_key_values", None),
+                    attention_mask=assistant_kwargs.get("attention_mask", None),
+                    decoder_attention_mask=assistant_kwargs.get("decoder_attention_mask", None),
+                    encoder_outputs=(assistant_kwargs.get("assistant_encoder_outputs", None),),
+                    past_key_values=assistant_kwargs.get("assistant_past_key_values", None),
                 )
-                if assistant_inputs.get("past_key_values", None) is not None:
-                    if assistant_model.config.is_encoder_decoder:
-                        input_ids_len = assistant_inputs["decoder_input_ids"].shape[-1]
-                    else:
-                        input_ids_len = assistant_inputs["input_ids"].shape[-1]
 
-                    if input_ids_len not in (1, 2):
-                        raise ValueError("The length of the input ids in assistant inputs should be 1 or 2")
+                # 1.2. check if the input ids length is correct
+                has_pkv = assistant_inputs.get("past_key_values", None) is not None
+                if has_pkv and assistant_inputs[input_ids_key].shape[-1] not in (1, 2):
+                    raise ValueError("The length of the input ids in assistant inputs should be 1 or 2")
 
+                # 1.3. use the assistant model to obtain the next candidate logits
                 assistant_model_outputs = assistant_model(**assistant_inputs)
 
-                # 1.2. greedily select the next candidate token
-                model_kwargs["assistant_past_key_values"] = assistant_model_outputs.past_key_values
+                # 1.4. greedily select the next candidate token
                 if len(logits_processor) > 0:
                     assistant_model_outputs.logits[:, -1, :] = logits_processor(
                         candidate_input_ids, assistant_model_outputs.logits[:, -1, :]
                     )
 
+                # 1.5. update assistant model inputs
```

Comment:
This "1.5" should start in `if assistant_kwargs ...`. The two lines below are still part of "greedily select the next candidate token"